### PR TITLE
cop: add abs rpn function

### PR DESCRIPTION
**Pull request by mapleFU**
## What have you changed? (mandatory)

Add implementation and tests for rpn function abs_int, abs_uint, abs_real, abs_decimal

If you want to see what was changed, you should go for `src/coprocessor/dag/rpn_expr` 

* mod.rs
* impl_math.rs

Was changed or created.

## What are the type of the changes? (mandatory)

* New feature (change which adds functionality)

## How has this PR been tested? (mandatory)

New unit tests.

### CODE REVIEW

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 1]
```diff
@@ -313,6 +313,67 @@ impl ArithmeticOp for DecimalMultiply {
     }
 }
 
+#[rpn_fn]
+#[inline]
+fn abs_int(arg: &Option<Int>) -> Result<Option<Int>> {
+    match arg {
+        Some(arg) => {
+            if *arg == std::i64::MIN {
```
Comment (kennytm) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: You could use `checked_abs` to avoid spelling out MIN explicitly. Also that `.to_owned` shouldn't be necessary

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 2]
```diff
@@ -672,4 +722,67 @@ mod tests {
...
+                .unwrap();
+            assert_eq!(output, expect_output, "{:?}, {:?}", arg, sig);
+        }
+    }
+
+    #[test]
+    fn test_abs_real() {
+        let test_cases: Vec<(ScalarFuncSig, Real, Option<Real>)> = vec![
+            (
+                ScalarFuncSig::AbsReal,
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: function signature is the same, no need to leave it here.

File: src/coprocessor/dag/rpn_expr/mod.rs [hunk 3]
```diff
@@ -170,6 +170,10 @@ fn map_pb_sig_to_rpn_func(value: ScalarFuncSig, children: &[Expr]) -> Result<Rpn
         ScalarFuncSig::IfNullTime => if_null_fn_meta::<DateTime>(),
         ScalarFuncSig::IfNullDuration => if_null_fn_meta::<Duration>(),
         ScalarFuncSig::IfNullJson => if_null_fn_meta::<Json>(),
+        ScalarFuncSig::AbsInt =>  abs_int_fn_meta(),
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/mod.rs]: Unnecessary whitespace after `=>`

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 4]
```diff
@@ -313,6 +313,56 @@ impl ArithmeticOp for DecimalMultiply {
...
+            None => Err(Error::overflow("BIGINT", &format!("abs({})", *arg)))?,
+            Some(arg_abs) => Ok(Some(arg_abs)),
+        },
+    }
+}
+
+#[rpn_fn]
+#[inline]
+fn abs_uint(arg: &Option<Int>) -> Result<Option<Int>> {
+    // TODO: make clear if it's ok
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: What's this TODO?

File: src/coprocessor/dag/rpn_expr/impl_arithmetic.rs [hunk 5]
```diff
@@ -313,6 +313,55 @@ impl ArithmeticOp for DecimalMultiply {
     }
 }
 
+#[rpn_fn]
+#[inline]
+fn abs_int(arg: &Option<Int>) -> Result<Option<Int>> {
```
Comment (rleungx) [inline src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: Do we need to move these functions into `impl_math.rs`?
/cc @breeswish

Comment (breeswish) [reply src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: Nice catch. Please move it to `impl_math` since it is in `builtin_math` in TiDB.

Comment (mapleFU) [reply src/coprocessor/dag/rpn_expr/impl_arithmetic.rs]: > Do we need to move these functions into `impl_math.rs`?
> /cc @breeswish

Done

File: src/coprocessor/dag/rpn_expr/impl_math.rs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    }
+}
+
+#[rpn_fn]
+#[inline]
+fn abs_real(arg: &Option<Real>) -> Result<Option<Real>> {
+    // abs returns NAN if the number is NAN, so don't worry about it
+    match arg {
+        Some(arg) => {
+            let f = arg.abs();
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_math.rs]: Shouldn't `abs()` already returns `Real`?

https://docs.rs/ordered-float/1.0.2/src/ordered_float/lib.rs.html#618

File: src/coprocessor/dag/rpn_expr/impl_math.rs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+#[rpn_fn]
+#[inline]
+fn abs_decimal(arg: &Option<Decimal>) -> Result<Option<Decimal>> {
+    use crate::coprocessor::codec::mysql::Res;
+
+    match arg {
+        Some(arg) => match arg.to_owned().abs() {
+            Res::Ok(v) => Ok(Some(v)),
+            Res::Truncated(_) => Err(Error::truncated())?,
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_math.rs]: Decimal's `Result` implements `Into` so that `Ok` becomes `Ok` and `Truncated` / `Overflow` becomes `Error`. I think we can utilize this trait here.

File: src/coprocessor/dag/rpn_expr/impl_math.rs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                .unwrap();
+            assert_eq!(output, expect_output, "{:?}, {:?}", arg, sig);
+        }
+    }
+
+    #[test]
+    fn test_abs_real() {
+        let test_cases: Vec<(ScalarFuncSig, Real, Option<Real>)> = vec![
+            (
+                ScalarFuncSig::AbsReal,
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_math.rs]: These function signatures are same. You can put it directly in `.evaluate(here)`

**Comment by breeswish** [comment]
Also please correctly sign-off your commits.

File: src/coprocessor/dag/rpn_expr/impl_math.rs [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+            None => Err(Error::overflow("BIGINT", &format!("abs({})", *arg)))?,
+            Some(arg_abs) => Ok(Some(arg_abs)),
+        },
+    }
+}
+
+#[rpn_fn]
+#[inline]
+fn abs_uint(arg: &Option<Int>) -> Result<Option<Int>> {
+    match arg {
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_math.rs]: Simply `Ok(*arg)` is enough?

**Comment by breeswish** [comment]
Please fix conflicts.

File: src/coprocessor/dag/rpn_expr/impl_math.rs [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            let arg = arg.parse::<Decimal>().ok();
+            let expect_output = expect_output.parse::<Decimal>().ok();
+            let output = RpnFnScalarEvaluator::new()
+                .push_param(arg.clone())
+                .evaluate(ScalarFuncSig::AbsDecimal)
+                .unwrap();
+            assert_eq!(
+                output,
+                expect_output,
+                "{:?} {:?}",
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_math.rs]: Looks like outputting one arg is enough. No need to output `ScalarFuncSig::AbsDecimal` since it can be inferred from the error stackdump.

**Comment by breeswish** [comment]
Please address remaining comments. Thanks!

File: src/coprocessor/dag/rpn_expr/impl_math.rs (hunk 7 above)
Comment (breeswish) [reply src/coprocessor/dag/rpn_expr/impl_math.rs]: You may refer to the original implementation of abs_decimal. It is not that complex.

**Comment by breeswish** [comment]
Also sign-off is missing for your latest commit 😅

File: src/coprocessor/dag/rpn_expr/impl_math.rs [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+    #[test]
+    fn test_abs_int() {
+        let test_cases = vec![
+            (ScalarFuncSig::AbsInt, -3, Some(3)),
+            (ScalarFuncSig::AbsInt, std::i64::MAX, Some(std::i64::MAX)),
+            (
+                ScalarFuncSig::AbsUInt,
+                std::u64::MAX as i64,
+                Some(std::u64::MAX as i64),
```
Comment (AndreMouche) [inline src/coprocessor/dag/rpn_expr/impl_math.rs]: How about the case which will cause overflow or truncated?

File: src/coprocessor/dag/rpn_expr/impl_math.rs [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+#[rpn_fn]
+#[inline]
+fn abs_real(arg: &Option<Real>) -> Result<Option<Real>> {
+    // abs returns NAN if the number is NAN, so don't worry about it
+    use num_traits::sign::Signed;
+
+    match arg {
+        Some(arg) => Ok(Some(<Real as Signed>::abs(arg))),
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_math.rs]: Simply:

```rust
fn abs_real(arg: &Option<Real>) -> Result<Option<Real>> {
    match arg {
        Some(arg) => Ok(Some(num_traits::Signed::abs(arg))),
        None => Ok(None),
    }
}
```

File: src/coprocessor/dag/rpn_expr/impl_math.rs [hunk 13]
```diff
@@ -26,25 +26,20 @@ fn abs_uint(arg: &Option<Int>) -> Result<Option<Int>> {
 #[inline]
 fn abs_real(arg: &Option<Real>) -> Result<Option<Real>> {
     // abs returns NAN if the number is NAN, so don't worry about it
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_math.rs]: Real would never be NAN. So this comment can be removed.

**Comment by sticnarf** [comment]
@mapleFU Add signed int overflow test case please.

File: src/coprocessor/dag/rpn_expr/impl_math.rs [hunk 14]
```diff
@@ -25,7 +25,7 @@ fn abs_uint(arg: &Option<Int>) -> Result<Option<Int>> {
 #[rpn_fn]
 #[inline]
 fn abs_real(arg: &Option<Real>) -> Result<Option<Real>> {
-    // abs returns NAN if the number is NAN, so don't worry about it
+    // arg is never nan
```
Comment (breeswish) [inline src/coprocessor/dag/rpn_expr/impl_math.rs]: So why not simply remove this comment 🤣 This comment provides no information, which looks to be useless.
